pkg/operator/admission/controller.go

```diff
@@ -0,0 +1,118 @@
+*/
+
+package admission
+
+import (
+	"encoding/json"
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"io/ioutil"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"net/http"
+)
+
+const (
+	jsonContentType = `application/json`
+)
+
+var (
+	universalDeserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+)
+
+// admitFunc is a callback for admission controller logic. Given an AdmissionRequest, it returns an error that will be shown when the operation is rejected.
+type admitFunc func(*v1beta1.AdmissionRequest, *clusterd.Context) error
+
+// doServeAdmitFunc parses the HTTP request for an admission controller webhook, and -- in case of a well-formed
+// request -- delegates the admission control logic to the given admitFunc. The response body is then returned as raw
+// bytes.
+func doServeAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) ([]byte, error) {
+	// Step 1: Request validation. Only handle POST requests with a body and json content type.
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, errors.Errorf("invalid method %q, only POST requess are allowed", r.Method)
+	}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.Wrap(err, "could not read request body")
+	}
+	if contentType := r.Header.Get("Content-Type"); contentType != jsonContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.Errorf("unsupported content type %q, only %q is supported", contentType, jsonContentType)
+	}
+	// Step 2: Parse the AdmissionReview request.
+	var admissionReviewReq v1beta1.AdmissionReview
+	if _, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.Wrap(err, "failed to deserialize request")
+	} else if admissionReviewReq.Request == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.New("malformed admission review. request is nil")
+	}
+	// Step 3: Construct the AdmissionReview response.
+	admissionReviewResponse := v1beta1.AdmissionReview{
+		Response: &v1beta1.AdmissionResponse{
+			UID: admissionReviewReq.Request.UID,
+		},
+	}
+	err = admit(admissionReviewReq.Request, rookContext)
+	if err != nil {
+		// If the handler returned an error, incorporate the error message into the response and deny the object
+		// creation.
+		admissionReviewResponse.Response.Allowed = false
+		admissionReviewResponse.Response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+	} else {
+		admissionReviewResponse.Response.Allowed = true
+	}
+	// Return the AdmissionReview with a response as JSON.
+	bytes, err := json.Marshal(&admissionReviewResponse)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal response")
```

Comment:
Would it be more clear as such?

```suggestion
		return nil, errors.Wrap(err, "failed to marshal response to AdmissionReview")
```